internal/clients/aws.go

```diff
@@ -2,17 +2,79 @@ package clients
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane-contrib/provider-tf-aws/apis/v1alpha1"
 )
 
-func ProviderConfigBuilder(ctx context.Context, client client.Client, mg xpresource.Managed) ([]byte, error) {
-	// TODO(hasan): Read ProviderConfig here and prepare a Terraform provider
-	//  config content with credentials accordingly.
+// ProviderConfigBuilder returns provider specific configuration like provider
+// credentials used to connect to cloud APIs in the expected form of a Terraform
+// provider.
+func ProviderConfigBuilder(ctx context.Context, client client.Client, mg resource.Managed) ([]byte, error) { //nolint:gocyclo
+	if mg.GetProviderConfigReference() == nil {
+		return nil, errors.New("no providerConfigRef provided")
+	}
+	pc := &v1alpha1.ProviderConfig{}
+	if err := client.Get(ctx, types.NamespacedName{Name: mg.GetProviderConfigReference().Name}, pc); err != nil {
+		return nil, errors.Wrap(err, "cannot get referenced Provider")
+	}
+
+	region, err := getRegion(mg)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get region")
+	}
+
+	t := resource.NewProviderConfigUsageTracker(client, &v1alpha1.ProviderConfigUsage{})
+	if err := t.Track(ctx, mg); err != nil {
+		return nil, errors.Wrap(err, "cannot track ProviderConfig usage")
+	}
+
+	var cfg *aws.Config
+	switch s := pc.Spec.Credentials.Source; s { //nolint:exhaustive
+	case xpv1.CredentialsSourceInjectedIdentity:
+		if cfg, err = xpawsclient.UsePodServiceAccount(ctx, []byte{}, xpawsclient.DefaultSection, region); err != nil {
+			return nil, errors.Wrap(err, "failed to use pod service account")
+		}
+	default:
+		data, err := resource.CommonCredentialExtractor(ctx, s, client, pc.Spec.Credentials.CommonCredentialSelectors)
+		if err != nil {
+			return nil, errors.Wrap(err, "cannot get credentials")
+		}
+		if cfg, err = xpawsclient.UseProviderSecret(ctx, data, xpawsclient.DefaultSection, region); err != nil {
+			return nil, errors.Wrap(err, "failed to use provider secret")
+		}
+	}
+	awsConf := xpawsclient.SetResolver(ctx, mg, cfg)
+	creds, err := awsConf.Credentials.Retrieve(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve aws credentials from aws config")
+	}
+
+	// TODO(hasan): figure out what other values could be possible set here.
+	//   e.g. what about setting an assume_role section: https://registry.terraform.io/providers/hashicorp/aws/latest/docs#argument-reference
+	tfPc := map[string]string{}
+	tfPc["region"] = awsConf.Region
+	tfPc["access_key"] = creds.AccessKeyID
+	tfPc["secret_key"] = creds.SecretAccessKey
+	tfPc["token"] = creds.SessionToken
+
+	return json.Marshal(tfPc)
+}
 
-	// TODO(hasan/muvaf) Terraform expects region or project information to be
-	//  in provider configuration different than Crossplane. Figure out how to
-	//  handle that.
-	return nil, nil
+func getRegion(obj runtime.Object) (string, error) {
```

Comment:
nitpick: since the operation consists of simple and self-explanatory calls, I'd suggest doing them directly in `ProviderConfigBuilder`

Reply:
I would prefer to keep it separate to make it more explicit that we are getting that field from somewhere else (i.e. spec of managed resource instead of provider config)